File: forAndroid/app/src/main/java/purpletreesoftware/karuahchess/model/gamerecord/GameRecordDataService.kt
/*
Karuah Chess is a chess playing program
Copyright (C) 2020 Karuah Software

Karuah Chess is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Karuah Chess is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

package purpletreesoftware.karuahchess.model.gamerecord

import android.content.ContentValues
import purpletreesoftware.karuahchess.common.App
import purpletreesoftware.karuahchess.database.DatabaseHelper
import purpletreesoftware.karuahchess.engine.KaruahChessEngineC
import java.lang.ref.WeakReference
import java.util.*
import kotlin.collections.ArrayList


@ExperimentalUnsignedTypes
class GameRecordDataService: IGameRecordDataService {

    private var _gameRecordDict: HashMap<Int, GameRecordArray>

    var currentGame: KaruahChessEngineC private set
    private var _tempBoardA: KaruahChessEngineC
    private var _tempBoardB: KaruahChessEngineC
    private var _tempBoardC: KaruahChessEngineC

    init {
        _objectInstances.add(WeakReference(this))
        _gameRecordDict = HashMap()
        currentGame = KaruahChessEngineC()
        _tempBoardA = KaruahChessEngineC()
        _tempBoardB = KaruahChessEngineC()
        _tempBoardC = KaruahChessEngineC()

        load()
    }

    /**
     * Loads game records
     */
    private fun load()
    {
        val board = KaruahChessEngineC()

        // Clear records from memory
        _gameRecordDict.clear()

        // Load game records in to memory
        val db = DatabaseHelper.getInstance(App.appContext).readableDatabase
        db.rawQuery("select * from GameRecord order by id", null).use { dbCursor ->
            while (dbCursor.moveToNext()) {
                val id = dbCursor.getInt(dbCursor.getColumnIndex("Id"))
                val boardSquareStr = dbCursor.getString(dbCursor.getColumnIndex("BoardSquareStr"))
                val gameStateStr = dbCursor.getString(dbCursor.getColumnIndex("GameStateStr"))
                board.setBoard(boardSquareStr)
                board.setState(gameStateStr)

                val recArray = GameRecordArray()
                recArray.id = id
                recArray.boardArray = board.getBoardArray()
                recArray.stateArray = board.getStateArray()
                _gameRecordDict.put(id, recArray)
            }
        }

        // If no records were loaded then create a default record
        if (_gameRecordDict.count() == 0)
        {
            reset()
        }

        // Set current game to latest bitboard
        val latestRecord = get()
        if(latestRecord != null) {
            currentGame.setBoardArray(latestRecord.boardArray)
            currentGame.setStateArray(latestRecord.stateArray)
        }

    }

    /**
     * Gets the current game as an array
     */
    override fun getCurrentGame(): GameRecordArray {
        val recArray = GameRecordArray()
        recArray.id = -1
        recArray.boardArray = currentGame.getBoardArray()
        recArray.stateArray = currentGame.getStateArray()
        return recArray
    }


    /**
     * Gets a record set by Id
     */
    override fun get(pId: Int): GameRecordArray? {
        if (_gameRecordDict.containsKey(pId)) {
            val rec = _gameRecordDict[pId]
            return rec
        }

        return null
    }


    /**
     * Gets the latest game record
     */
    override fun get(): GameRecordArray? {
        val latestId = getMaxId()
        return get(latestId)
    }

    /**
     * Record the current state of the game as a record
     */
    override fun recordGameState(pWhiteClockOffset: Int, pBlackClockOffset: Int): Long
    {
        var result = 0L

        // Set the clocks
        currentGame.setStateWhiteClockOffset(pWhiteClockOffset)
        currentGame.setStateBlackClockOffset(pBlackClockOffset)

        // create game record
        val nextId = getMaxId() + 1
        val boardSquareStr = currentGame.getBoard()
        val gameStateStr = currentGame.getState()

        val gameRecordArray = GameRecordArray()
        gameRecordArray.id = nextId
        gameRecordArray.boardArray = currentGame.getBoardArray()
        gameRecordArray.stateArray = currentGame.getStateArray()

        if (!_gameRecordDict.containsKey(gameRecordArray.id)) {

            // Add record to dictionary
            _gameRecordDict.put(gameRecordArray.id, gameRecordArray)

            // Add record to database
            val db = DatabaseHelper.getInstance(App.appContext).writableDatabase
            val contentValues = ContentValues()
            contentValues.put("Id", nextId)
            contentValues.put("BoardSquareStr", boardSquareStr)
            contentValues.put("GameStateStr", gameStateStr)

            result = db.insert("GameRecord", null, contentValues)

        }

        return result
    }

    /**
     * Updates a board record
     */
    override fun updateGameState(pGameRecordArray: GameRecordArray): Int
    {
        var result = 0

        if (_gameRecordDict.containsKey(pGameRecordArray.id))
        {
            _tempBoardC.setBoardArray(pGameRecordArray.boardArray)
            _tempBoardC.setStateArray(pGameRecordArray.stateArray)
            val boardSquareStr: String = _tempBoardC.getBoard()
            val gameStateStr: String = _tempBoardC.getState()

            // Add record to dictionary
            _gameRecordDict[pGameRecordArray.id] = pGameRecordArray

            // Add record to database
            val db = DatabaseHelper.getInstance(App.appContext).writableDatabase
            val contentValues = ContentValues()
            contentValues.put("Id", pGameRecordArray.id)
            contentValues.put("BoardSquareStr", boardSquareStr)
            contentValues.put("GameStateStr", gameStateStr)
            result = db.update("GameRecord", contentValues, "Id=?", arrayOf(pGameRecordArray.id.toString()))


            // Update the current game if updating max record
            val maxId = getMaxId()
            if (pGameRecordArray.id == maxId)
            {
                currentGame.setBoardArray(pGameRecordArray.boardArray);
                currentGame.setStateArray(pGameRecordArray.stateArray);
            }
        }

        return result
    }


    /**
     * Clears game record
     */
    override fun reset() {


        val db = DatabaseHelper.getInstance(App.appContext).writableDatabase

        db.delete("GameRecord", null, null)

        // Clear dictionary
        _gameRecordDict.clear()

        // Reset the current game
        currentGame.reset()

        // Create record of default setup
        recordGameState(0,0)
    }

    /**
     * Clear from a given [pId]
     */
    private fun clearFrom(pId: Int)
    {


        val db = DatabaseHelper.getInstance(App.appContext).writableDatabase

        db.delete("GameRecord", "Id >= ?", arrayOf(pId.toString()))


        // Remove from dictionary
        for (key in _gameRecordDict.keys.toList())
        {
            if (key >= pId) _gameRecordDict.remove(key)
        }

    }


    /**
     * Undo last move
     */
    override fun undo(): Boolean {
        val lastMoveId = getMaxId()

        val previousBoard = get(lastMoveId - 1)
        var returnValue = false

        if (previousBoard != null) {
            // Remove record from database
            clearFrom(lastMoveId)

            // Set the current game to a previous board
            currentGame.setBoardArray(previousBoard.boardArray);
            currentGame.setStateArray(previousBoard.stateArray);

            // Return true
            returnValue = true
        }

        return returnValue
    }

    /**
     * Compares two boards and returns sq ids that have changed
     */
    override fun getBoardSquareChanges(pBoardA: GameRecordArray?, pBoardB: GameRecordArray?): ULong {

        var mask: ULong = 0b10000000_00000000_00000000_00000000_00000000_00000000_00000000_00000000uL
        var changedIndexes = 0UL

        //Loop through bit boards and detectchanges
        if (pBoardA != null && pBoardB != null) {
            _tempBoardA.setBoardArray(pBoardA.boardArray);
            _tempBoardA.setStateArray(pBoardA.stateArray);
            _tempBoardB.setBoardArray(pBoardB.boardArray);
            _tempBoardB.setStateArray(pBoardB.stateArray);

            for (i in 0..63) {
                val sqAspin = _tempBoardA.getSpin(i)
                val sqBspin = _tempBoardB.getSpin(i)
                if (sqAspin != sqBspin) changedIndexes = changedIndexes or mask
                mask = mask shr 1
            }
        }
        return changedIndexes
    }


    /**
     * Gets last id in the table
     */
    override fun getMaxId(): Int
    {
        var maxId = 0

        val db = DatabaseHelper.getInstance(App.appContext).readableDatabase
        db.rawQuery("select max(Id) from GameRecord", null).use { dbCursor ->
            if (dbCursor.count > 0) {
                dbCursor.moveToFirst()
                maxId = dbCursor.getInt(0)
            }
        }

        return maxId

    }


    /**
     * Get all record Ids as an ArrayList
     */
    override fun getAllRecordIDList(): ArrayList<Int>
    {

        val recordIDList = ArrayList<Int>()

        // Get list of Ids from the database
        val db = DatabaseHelper.getInstance(App.appContext).readableDatabase
        db.rawQuery("select id from GameRecord order by id", null).use { dbCursor ->
            while (dbCursor.moveToNext()) {
                val Id = dbCursor.getInt(0)
                recordIDList.add(Id)
            }
        }

        return recordIDList

    }


    companion object {
        private var _objectInstances = HashSet<WeakReference<GameRecordDataService>>()

        /**
         * Refreshes all instances
         */

        fun reloadAllInstances() {
            var obj: GameRecordDataService?
            val objectInstancesValid = HashSet<WeakReference<GameRecordDataService>>()

            for (weakRef in _objectInstances) {
                obj = weakRef.get()
                if (obj != null) {
                    obj.load()
                    objectInstancesValid.add(weakRef)
                }
            }

            _objectInstances = objectInstancesValid
        }

    }


}

File: forWinUI/KaruahChess/Model/gamerecord/GameRecordDataService.cs
/*
Karuah Chess is a chess playing program
Copyright (C) 2020-2023 Karuah Software

Karuah Chess is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Karuah Chess is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

using KaruahChess.Database;
using System;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Data.Sqlite;
using KaruahChessEngine;



namespace KaruahChess.Model
{
    public class GameRecordDataService : IGameRecordDataService
    {

        Dictionary<int, GameRecordArray> _gameRecordDict;
        KaruahChessEngineClass _tempBoardA;
        KaruahChessEngineClass _tempBoardB;
        KaruahChessEngineClass _tempBoardC;

        public KaruahChessEngineClass CurrentGame { get; private set; }
        public long transactionId {get; private set;} 

        // Constructor
        private GameRecordDataService()
        {          
           
          
            _gameRecordDict = new Dictionary<int, GameRecordArray>();

            CurrentGame = new KaruahChessEngineClass();            
            _tempBoardA = new KaruahChessEngineClass();
            _tempBoardB = new KaruahChessEngineClass();
            _tempBoardC = new KaruahChessEngineClass();

            transactionId = 0;

            Load();
        }

        /// <summary>
        /// Singleton accessor
        /// </summary>
        public static GameRecordDataService instance { get; private set; } = new GameRecordDataService();


        /// <summary>
        /// Loads parameters
        /// </summary>
        public void Load()
        {
            KaruahChessEngineClass board = new KaruahChessEngineClass();

            // Clear records from memory
            _gameRecordDict.Clear();

            // Load game records in to memory            
            var connection = KaruahChessDB.GetDBConnection();            
            using (var command = connection.CreateCommand())
            {
                command.CommandText = $"select * from {KaruahChessDB.GameRecordTableName} order by id";
                SqliteDataReader reader = command.ExecuteReader();

               
                    while (reader.Read())
                    {
                        int Id = Convert.ToInt32(reader["Id"]);
                        String boardSquareStr = Convert.ToString(reader["BoardSquareStr"]);
                        String gameStateStr = Convert.ToString(reader["GameStateStr"]);
                        String moveSANStr = Convert.ToString(reader["MoveSANStr"]);
                        board.SetBoard(boardSquareStr);
                        board.SetState(gameStateStr);

                        GameRecordArray recArray = new GameRecordArray();
                        recArray.Id = Id;
                        board.GetBoardArray(recArray.BoardArray);
                        board.GetStateArray(recArray.StateArray);
                        recArray.MoveSAN = moveSANStr;

                        _gameRecordDict.Add(Id, recArray);
                    }
                

            }

            connection.Close();

            
      
            // If no records were loaded then create a default record
            if (_gameRecordDict.Count == 0)
            {
                Reset(0, 0);
            }

            // Set current game to latest bitboard
            var latestRecord = Get();
            CurrentGame.SetBoardArray(latestRecord.BoardArray);
            CurrentGame.SetStateArray(latestRecord.StateArray);
                        
        }

        /// <summary>
        /// Gets the current game as an array
        /// </summary>
        /// <param name="pId"></param>
        /// <returns></returns>
        public GameRecordArray GetCurrentGame()
        {
            GameRecordArray recArray = new GameRecordArray();
            recArray.Id = -1;
            CurrentGame.GetBoardArray(recArray.BoardArray);
            CurrentGame.GetStateArray(recArray.StateArray);

            return recArray;

        }

        /// <summary>
        /// Gets a record set by Id
        /// </summary>
        /// <param name="pId"></param>
        /// <returns></returns>
        public GameRecordArray Get(int pId)
        {
            if (_gameRecordDict.ContainsKey(pId)) {
                GameRecordArray rec = _gameRecordDict[pId];                
                return rec;
            }
            else {
                return null;
            }
           
        }

        /// <summary>
        /// Gets the latest game record
        /// </summary>
        /// <returns></returns>
        public GameRecordArray Get()
        {
            int latestId = GetMaxId();
            return (Get(latestId));

        }

        /// <summary>
        /// Record the current state of the game as a record
        /// </summary>
        /// <returns></returns>
        public int RecordGameState(int pWhiteClockOffset, int pBlackClockOffset, string pMoveSAN)
        {
            int result = 0;

            // Set the clocks
            CurrentGame.SetStateWhiteClockOffset(pWhiteClockOffset);
            CurrentGame.SetStateBlackClockOffset(pBlackClockOffset);


            // create game record            
            int nextId = GetMaxId() + 1;
            String boardSquareStr = CurrentGame.GetBoard();
            String gameStateStr = CurrentGame.GetState();

            var gameRecordArray = new GameRecordArray();
            gameRecordArray.Id = nextId;
            CurrentGame.GetBoardArray(gameRecordArray.BoardArray);
            CurrentGame.GetStateArray(gameRecordArray.StateArray);
            gameRecordArray.MoveSAN = pMoveSAN;

            if (!_gameRecordDict.ContainsKey(gameRecordArray.Id)) {

                // Add record to dictionary
                _gameRecordDict.Add(gameRecordArray.Id, gameRecordArray);

                // Add record to database
                var connection = KaruahChessDB.GetDBConnection();                
                    using (var command = connection.CreateCommand())
                    {
                        command.CommandText = $"INSERT INTO {KaruahChessDB.GameRecordTableName} (Id, BoardSquareStr, GameStateStr, MoveSANStr) Values (@Id, @BoardSquareStr, @GameStateStr, @MoveSANStr);";                    
                        command.Parameters.Add(new SqliteParameter("@Id", nextId));
                        command.Parameters.Add(new SqliteParameter("@BoardSquareStr", boardSquareStr));
                        command.Parameters.Add(new SqliteParameter("@GameStateStr", gameStateStr));                    
                        command.Parameters.Add(new SqliteParameter("@MoveSANStr", gameRecordArray.MoveSAN));
                        result = command.ExecuteNonQuery();
                        transactionId++;                
                    }

                    connection.Close();
                }
            

            return result;
        }

        /// <summary>
        /// Updates a board record
        /// </summary>
        /// <returns></returns>
        public int UpdateGameState(GameRecordArray pGameRecordArray)
        {
            int result = 0;
            
            if (_gameRecordDict.ContainsKey(pGameRecordArray.Id))
            {
                _tempBoardC.SetBoardArray(pGameRecordArray.BoardArray);
                _tempBoardC.SetStateArray(pGameRecordArray.StateArray);
                string boardSquareStr = _tempBoardC.GetBoard();
                string gameStateStr = _tempBoardC.GetState();

                // Add record to dictionary
                _gameRecordDict[pGameRecordArray.Id] = pGameRecordArray;
                
                // Single DB connection for both updates
                var connection = KaruahChessDB.GetDBConnection();
                using (var command = connection.CreateCommand())
                {
                    // Update current record (board/state + cleared SAN)
                    command.CommandText = $"Update {KaruahChessDB.GameRecordTableName} set BoardSquareStr=@BoardSquareStr, GameStateStr=@GameStateStr, MoveSANStr=@MoveSANStr where id=@Id;";
                    command.Parameters.Add(new SqliteParameter("@Id", pGameRecordArray.Id));
                    command.Parameters.Add(new SqliteParameter("@BoardSquareStr", boardSquareStr));
                    command.Parameters.Add(new SqliteParameter("@GameStateStr", gameStateStr));
                    command.Parameters.Add(new SqliteParameter("@MoveSANStr", string.Empty));
                    result = command.ExecuteNonQuery();
                    transactionId++;                    
                }
                connection.Close();

                // Update the current game if updating max record
                var maxId = GetMaxId();
                if (pGameRecordArray.Id == maxId)
                {
                    CurrentGame.SetBoardArray(pGameRecordArray.BoardArray);
                    CurrentGame.SetStateArray(pGameRecordArray.StateArray);
                }

            }


            return result;
        }

        /// <summary>
        /// Clears game record
        /// </summary>
        public void Reset(int pWhiteClockOffset, int pBlackClockOffset) {

            // Clear records
           using (var connection = KaruahChessDB.GetDBConnection()) {   
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = $"Delete from {KaruahChessDB.GameRecordTableName};";                    
                    command.ExecuteNonQuery();
                    transactionId++;
                }
                connection.Close();
            }
           
            // Clear dictionary
            _gameRecordDict.Clear();

            // Reset the current game
            CurrentGame.Reset();

            // Create record of default setup
            RecordGameState(pWhiteClockOffset, pBlackClockOffset, string.Empty);
        }

        /// <summary>
        /// Clear start
        /// </summary>
        /// <param name="pId"></param>
        private void ClearFrom(int pId)
        {
            using (var connection = KaruahChessDB.GetDBConnection())
            {
                // Remove from database                
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = $"Delete from {KaruahChessDB.GameRecordTableName} where id >= @Id;";
                    command.Parameters.Add(new SqliteParameter("Id", pId));                    
                    command.ExecuteNonQuery();
                    transactionId++;
                }

                connection.Close();
            }

            // Remove from dictionary
            foreach (var key in _gameRecordDict.Keys.ToList())
            {
                if (key >= pId) _gameRecordDict.Remove(key);
            }
                       
        }

        

        /// <summary>
        /// Undo last move
        /// </summary>
        /// <param name="pBoardSquareDS"></param>
        /// <returns></returns>
        public bool Undo()
        {
            var lastMoveId = GetMaxId();
            
            GameRecordArray previousBoard = Get(lastMoveId - 1);
            bool returnValue = false;

            if (previousBoard != null)
            { 
                // Remove record from database
                ClearFrom(lastMoveId);

                // Set the current game to a previous board
                CurrentGame.SetBoardArray(previousBoard.BoardArray);
                CurrentGame.SetStateArray(previousBoard.StateArray);

                // Return true
                returnValue = true;
            }


            return returnValue;
        }

       
        /// <summary>
        /// Compares two boards and returns sq ids that have changed
        /// </summary>
        /// <returns></returns>
        public HashSet<int> GetBoardSquareChanges(GameRecordArray pBoardA, GameRecordArray pBoardB)
        {
            HashSet<int> changedIndexes = new HashSet<int>();                       

            //Loop through bit boards and detectchanges
            if (pBoardA != null && pBoardB != null) {
                _tempBoardA.SetBoardArray(pBoardA.BoardArray);
                _tempBoardA.SetStateArray(pBoardA.StateArray);
                _tempBoardB.SetBoardArray(pBoardB.BoardArray);
                _tempBoardB.SetStateArray(pBoardB.StateArray);

                for (int i = 0; i < 64; i++)
                {
                    int sqAspin = _tempBoardA.GetSpin(i);
                    int sqBspin = _tempBoardB.GetSpin(i);
                    if(sqAspin != sqBspin) changedIndexes.Add(i);

                }
            }
            return changedIndexes;
        }

        
        /// <summary>
        /// Gets last id in the table
        /// </summary>
        /// <returns></returns>
        public int GetMaxId()
        {
            int maxId = 0;

            using (var connection = KaruahChessDB.GetDBConnection())
            {   
                // Remove from database                
                using (var command = connection.CreateCommand())
                {
                    command.CommandText = $"select max(Id) from {KaruahChessDB.GameRecordTableName};";                    
                    Object result = command.ExecuteScalar();
                    if (result != null && result != DBNull.Value) maxId = Convert.ToInt32(result);

                }

                connection.Close();            
            }

            return maxId;

        }

        /// <summary>
        /// Generates a new transacation id
        /// </summary>
        public void newTransaction()
        {
            transactionId++;
        }

        /// <summary>
        /// Gets all record ids
        /// </summary>
        /// <returns></returns>
        public List<int> GetAllRecordIDList()
        {

            List<int> recordIDList = new List<int>();            

            // Get list of Ids from the database
            var connection = KaruahChessDB.GetDBConnection();
            using (var command = connection.CreateCommand())
            {
                command.CommandText = $"select id from {KaruahChessDB.GameRecordTableName} order by id";
                SqliteDataReader reader = command.ExecuteReader();

                while (reader.Read())
                {
                    recordIDList.Add(Convert.ToInt32(reader["Id"]));
                }

            }

            connection.Close();

            return recordIDList;

        }

        /// <summary>
        /// Returns a count of the records in the dictionary
        /// </summary>
        public int RecordCount() {
            return _gameRecordDict.Count();
        }

        /// <summary>
        /// Gets the game history
        /// </summary>        
        public SortedList<int, GameRecordArray> GameHistory()
        {            
            SortedList<int, GameRecordArray> history = new SortedList<int, GameRecordArray>();

            foreach (var kvp in _gameRecordDict)
            {
                history.Add(kvp.Key, kvp.Value);
            }

            return history;
        }


        /// <summary>
        /// Determines the active move color for a specific record
        /// </summary>  
        public int GetActiveMoveColour(int pId)
        {
            if (_gameRecordDict.ContainsKey(pId))
            {
                GameRecordArray rec = _gameRecordDict[pId];
                return rec.StateArray[0];
            }
            else
            {
                return 0;
            }           
        }



        /// <summary>
        /// Get the game status for a specific record
        /// </summary>  
        public int GetStateGameStatus(int pId)
        {
            if (_gameRecordDict.ContainsKey(pId))
            {
                GameRecordArray rec = _gameRecordDict[pId];
                return rec.StateArray[5];
            }
            else
            {
                return -1;
            }
        }

    }
}
